python/ray/util/sgd/torch/worker_group.py

```diff
@@ -0,0 +1,480 @@
+import io
+import logging
+import time
+from datetime import timedelta
+
+import ray
+import torch
+from ray.exceptions import RayActorError
+from ray.util.sgd.torch.distributed_torch_runner import \
+    LocalDistributedRunner, DistributedTorchRunner
+from ray.util.sgd.torch.torch_runner import TorchRunner
+from ray.util.sgd.utils import check_for_failure
+
+RESIZE_COOLDOWN_S = 10
+logger = logging.getLogger(__name__)
+
+
+class BaseWorkerGroup:
+    def start_workers(self, num_workers, params, dist_params,
+                      initialization_hook, timeout_s, num_cpus_per_worker,
+                      use_gpu):
+        raise NotImplementedError
+
+    def apply_all_operators(self, fn):
+        raise NotImplementedError
+
+    def apply_all_workers(self, fn):
+        raise NotImplementedError
+
+    def get_local_operator(self):
+        raise NotImplementedError
+
+    def get_model(self):
+        raise NotImplementedError
+
+    def load_state_dict(self, state_dict, blocking=False):
+        raise NotImplementedError
+
+    def shutdown(self, force=False):
+        raise NotImplementedError
+
+    def state_dict(self, to_cpu=False):
+        raise NotImplementedError
+
+    def train(self, num_steps=None, profile=False, info=None, dataset=None):
+        raise NotImplementedError
+
+    def validate(self, num_steps=None, profile=False, info=None):
+        raise NotImplementedError
+
+    def should_resize(self):
+        raise NotImplementedError
+
+    def resize_workers(self, max_retries=10):
+        raise NotImplementedError
+
+    def get_num_workers(self):
+        raise NotImplementedError
+
+
+class RemoteWorkerGroup(BaseWorkerGroup):
+    """Groups together multiple DistributedTorchRunner workers."""
+
+    def __init__(self):
+        self.remote_workers = []
+        self.num_workers = 0
+        self._last_resize = float("-inf")
+
+    def _init_workers(self, num_workers, params, num_cpus_per_worker, use_gpu):
+        self.num_workers = num_workers
+        self.num_cpus_per_worker = num_cpus_per_worker
+        self.use_gpu = use_gpu
```

Comment:
are these properties? can you make it so that the setting of these is explicit?

Reply:
if they are properties (i.e, externally accessible) can you document?

Reply:
actually maybe you can leave the explicit setting out.

Reply:
I cleaned this up a lot. All state is passed in through __init__ and shouldn't be changed again afterwards.